Add proper check mode support to the script module

samdoran opened the pull request:
##### SUMMARY

The `script` action plugin actually ran scripts when in check mode since the default arguments for action plugins sets `self._supports_check_mode = True` and there was nothing in the `script` action plugin to prevent the script from actually running on the managed host(s).

Fixes #30676 

Added a bunch of tests to make sure no changes are made when run in check mode.

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->

 - Bugfix Pull Request


##### COMPONENT NAME
`lib/ansible/plugins/action/script.py`

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
2.5
```


##### ADDITIONAL INFORMATION

This has been the case in all Ansible version back to 1.9 according to the issue. There may be a number of people inadvertently relying on this behavior.

samdoran committed:
Do not run script in check mode

Fixes #30676

samdoran committed:
Reformat script integration test

samdoran committed:
Add integration tests for check mode of script module

nitzmahone commented on test/integration/targets/script/tasks/main.yml:
```diff
@@ -81,5 +126,73 @@
+    var: _afile_stat
+    verbosity: 2
+
+- name: Assert that a change was reported but the script did make changes
```
s/did/did not/

nitzmahone reviewed (changes requested):
Looks good other than one little comment fix. Can you also add the check mode stuff to the `win_script` test target as well (uses the same action impl, but different tests)?

bcoca commented on lib/ansible/plugins/action/script.py:
```diff
@@ -96,6 +98,10 @@ def run(self, tmp=None, task_vars=None):
+
+        if self._play_context.check_mode:
+            return dict(changed=True)
+
```
Not sure you need this down here, if you just set `self._supports_check_mode = False` in `__init__` or in `run` before calling `super` the base class will already skip this task. In any case you already have `result` dict with other information you might want to return `result['changed'] = True; return result`

samdoran replied:
I like doing it down here because it allows the module to support check mode, not just be skipped. That means you can run it in check mode and if there are `creates:` or `removes:` parameters, those get evaluated in check mode. Otherwise, it'll just report a change, like `script` always does by default, but not actually run the script.

I'll update the `results` dict rather than returning a new one.

samdoran committed:
Fix name on test

samdoran committed:
Cleanup temp file

samdoran committed:
win_script integration test syntaxt changes

samdoran committed:
Add check mode tests for win_script

mattclay commented:
CI failure in the win_script integration tests: https://app.shippable.com/github/ansible/ansible/runs/42056/7/tests

samdoran committed:
Use proper variable in test

samdoran commented:
@bcoca @nitzmahone Can you look at this again? I've added Windows tests and have it working properly in check for both Windows and Linux.

bcoca commented on lib/ansible/plugins/action/script.py:
```diff
@@ -76,6 +76,8 @@ def run(self, tmp=None, task_vars=None):
         # and everything else is an argument to the script which we need later
         # to append to the remote command
+        # FIXME: If the path to the script has a space in it, this silently fails to do anything
+        # This code makes a bad assumption by splitting on spaces.
```
we should use 'shell splitting' here ... shlex .... but that is aside from this PR

samdoran replied:
Noted. Thanks for the info.

bcoca commented:
Not sure the check mode is in the correct place, we should avoid transfering the script file since there is no 'dry run' mode for shell.

samdoran commented:
@bcoca So does that mean you think it’s better to just disable check mode altogether for the script module? That would be much easier, though I like having it properly report changes if `args` are provided.

bcoca commented:
its worth at least verifying that we can access the script itself

samdoran committed:
Fail if source file does not exist

samdoran committed:
Verify script is accessible and don't copy in check mode

Use shlex to properly split shell arguments, though a path with spaces in it still needs to be quoted in the playbook.
Add note to docs describing such.
Improve error message if file is not found indicating there may be a space in the path.

samdoran committed:
Properly encode path now that path is split using shlex

samdoran committed:
Allow for spaces in both path and script name

samdoran committed:
Add unicode character test to Linux script tests

samdoran committed:
Add Linux test for space in path to script

nitzmahone commented:
Test failure is an unrelated transient networking issue- merging.
